node/network/collator-protocol/src/collator_side.rs

```diff
@@ -240,25 +241,86 @@ where
 /// Issue a connection request to a set of validators.
 async fn connect_to_validators<Context>(
 	ctx: &mut Context,
+	relay_parent: Hash,
 	state: &mut State,
 	validators: Vec<ValidatorId>,
 ) -> Result<()>
 where
 	Context: SubsystemContext<Message = CollatorProtocolMessage>
 {
+	let (result, revoke) = connect_to_validators_impl(ctx, relay_parent, validators).await?;
+
+	for (validator_id, peer_id) in result.into_iter() {
+		state.known_validators.insert(peer_id, validator_id);
+	}
+
+	// TODO: do we want to revoke requests at all?
+	drop(revoke);
+	Ok(())
+}
+
+type RevokeConnectionRequest = oneshot::Sender<()>;
+
+// TODO: make this code reusable by other subsystems
+// put in subsystem-util?
+async fn connect_to_validators_impl<Context: SubsystemContext>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+	validators: Vec<ValidatorId>,
+) -> Result<(Vec<(ValidatorId, PeerId)>, RevokeConnectionRequest)> {
+	// ValidatorId -> AuthorityDiscoveryId
 	let (tx, rx) = oneshot::channel();
 
-	ctx.send_message(AllMessages::NetworkBridge(
-		NetworkBridgeMessage::ConnectToValidators(PeerSet::Collation, validators, tx),
+	ctx.send_message(AllMessages::RuntimeApi(
+		RuntimeApiMessage::Request(
+			relay_parent,
+			RuntimeApiRequest::ValidatorDiscovery(validators.clone(), tx),
+		)
 	)).await?;
 
-	let mut validators_ids = rx.await?;
+	let authorities = rx.await??;
 
-	for id in validators_ids.drain(..) {
-		state.known_validators.insert(id.1, id.0);
+	let mut validator_map = validators.into_iter()
+		.zip(authorities.into_iter())
+		.filter_map(|(k, v)| v.map(|v| (v, k)))
+		.collect::<HashMap<AuthorityDiscoveryId, ValidatorId>>();
+
+	let authorities: Vec<_> = validator_map.keys().cloned().collect();
+
+	let (mut rx, revoke) = connect_to_authorities(ctx, authorities).await?;
+
+	// TODO: wait for only one PeerId maybe and for the rest in the background?
+	// or use a timeout?
```

Comment:
Yes, I don't think we should stall here if we cannot connect to anyone

Reply:
Hmm, I'm confused now given that the collator handles `PeerConnected` and `PeerDisconnected` events.
`ConnectToValidators` returns the list when the validators are already connected. Is this not what we want?

Otherwie, there is a problem with a race condition if we receive `PeerConnected` message from the `NetworkBridge` before `ConnectToValidators` response.